src/libmongoc/tests/test-mongoc-secure-channel.c

```diff
@@ -0,0 +1,263 @@
+      ASSERT_OR_PRINT (ok, error);
+      // Use again.
+      ok = connect_with_secure_channel_cred (&ssl_opt, cred_ptr, &error);
+      ASSERT_OR_PRINT (ok, error);
+      mongoc_shared_ptr_reset_null (&cred_ptr);
+   }
+}
+
+typedef struct {
+   size_t failures;
+   size_t failures2;
+} cert_failures;
+
+void
+count_cert_failures (mongoc_log_level_t log_level, const char *log_domain, const char *message, void *user_data)
+{
+   cert_failures *cf = user_data;
+   if (strstr (message, "Failed to open file: 'does-not-exist.pem'")) {
+      cf->failures++;
+   }
+   if (strstr (message, "Failed to open file: 'does-not-exist-2.pem'")) {
+      cf->failures2++;
+   }
+}
+
+static bool
+try_ping (mongoc_client_t *client, bson_error_t *error)
+{
+   return mongoc_client_command_simple (client, "admin", tmp_bson (BSON_STR ({"ping" : 1})), NULL, NULL, error);
+}
+
+static bool
+try_ping_with_reconnect (mongoc_client_t *client, bson_error_t *error)
+{
+   // Force a connection error with a failpoint:
+   if (!mongoc_client_command_simple (client,
+                                      "admin",
+                                      tmp_bson (BSON_STR ({
+                                         "configureFailPoint" : "failCommand",
+                                         "mode" : {"times" : 1},
+                                         "data" : {"closeConnection" : true, "failCommands" : ["ping"]}
+                                      })),
+                                      NULL,
+                                      NULL,
+                                      error)) {
+      return false;
+   }
+
+   // Expect first ping to fail:
+   if (try_ping (client, error)) {
+      bson_set_error (error, 0, 0, "unexpected: ping succeeded, but expected to fail");
+      return false;
+   }
+
+   // Ping again:
+   return try_ping (client, error);
+}
+
+static void
+test_secure_channel_shared_creds_client (void *unused)
+{
+   BSON_UNUSED (unused);
+
+   bson_error_t error;
+
+   // Save log function:
+   mongoc_log_func_t saved_log_func;
+   void *saved_log_data;
+   _mongoc_log_get_handler (&saved_log_func, &saved_log_data);
+
+   // Set log function to count failed attempts to load client cert:
+   cert_failures cf = {0};
+   mongoc_log_set_handler (count_cert_failures, &cf);
+
+   // Test client:
+   {
+      mongoc_client_t *client = test_framework_new_default_client ();
+
+      // Set client cert to a bad path:
+      {
+         mongoc_ssl_opt_t ssl_opt = *test_framework_get_ssl_opts ();
+         ssl_opt.pem_file = "does-not-exist.pem";
+         mongoc_client_set_ssl_opts (client, &ssl_opt);
+      }
+
+      // Expect insert OK. Cert fails to load, but server configured with --tlsAllowConnectionsWithoutCertificates:
+      {
+         bool ok = try_ping (client, &error);
+         ASSERT_OR_PRINT (ok, error);
+      }
+
+      // Expect exactly one attempt to load the client cert:
+      ASSERT_CMPSIZE_T (1, ==, cf.failures);
```

Comment:
Can use the type-generic `mlib_check` macros

```suggestion
      mlib_check (1, eq, cf.failures);
```